operator/beiboot/clusterstate.py

```diff
@@ -0,0 +1,277 @@
+import base64
+import random
+import uuid
+from asyncio import sleep
+from datetime import datetime
+
+import kubernetes as k8s
+import kopf
+import yaml
+from statemachine import State
+
+from beiboot.configuration import BeibootConfiguration, ClusterConfiguration
+from beiboot.provider.abstract import AbstractClusterProvider
+from beiboot.provider.factory import cluster_factory, ProviderType
+from beiboot.resources.services import ports_to_services, gefyra_service
+from beiboot.resources.utils import handle_create_namespace, handle_create_service
+from beiboot.utils import StateMachine, AsyncState
+
+
+class BeibootCluster(StateMachine):
+    """
+    A Beiboot cluster is implemented as a state machine
+    The body of the Beiboot CRD is available as self.model
+    """
+
+    requested = State("Cluster requested", initial=True, value="REQUESTED")
+    creating = AsyncState("Cluster creating", value="CREATING")
+    pending = AsyncState("Cluster pending", value="PENDING")
+    running = AsyncState("Cluster running", value="RUNNING")
+    ready = AsyncState("Cluster running", value="READY")
+    error = AsyncState("Cluster error", value="ERROR")
+    terminating = AsyncState("Cluster terminating", value="TERMINATING")
```

Comment:
maybe "terminated", as this is the last state and the termination should be finished, or is this the state before reaching "terminated"?

Reply:
"Terminated" means this Beiboot cluster (CRD) does not exist anymore in K8s